File: Example/VISPER/Example3/ExampleFeature3.m
//
//  ExampleFeature3.m
//  VISPER
//
//  Created by Bartel on 20.08.15.
//  Copyright (c) 2015 Jan Bartel. All rights reserved.
//

#import "ExampleFeature3.h"

@implementation ExampleFeature3

-(instancetype)init{
    self = [super init];
    if(self){
        [self addRoutePattern:@"/example3"];
    }
    return self;
}

-(UIViewController*)controllerForRoute:(NSString*)routePattern
                        routingOptions:(NSObject<IVISPERRoutingOption>*)options
                        withParameters:(NSDictionary*)parameters{
    
    if ([routePattern isEqualToString:@"/example3"]) {
        Example3VisperViewControllerPresenter *example3VCPresenter =
        [[Example3VisperViewControllerPresenter alloc] initWithWireframe:self.wireframe];
        
        Example3VisperViewController *viewController =
        [[Example3VisperViewController alloc] initWithNibName:@"Example3VisperViewController"
                                                       bundle:nil];
        [viewController addVisperPresenter:example3VCPresenter];
        
        return viewController;
    }

    
    
    return nil;
}


-(NSObject<IVISPERRoutingOption> *)optionForRoutePattern:(NSString *)routePattern
                                              parameters:(NSDictionary*)parameters
                                          currentOptions:(NSObject<IVISPERRoutingOption>*)currentOptions{
    if(currentOptions){
        return currentOptions;
    }
    
    if ([routePattern isEqualToString:@"/example2"]) {
        return [self.wireframe pushRoutingOption:YES];
    }
    
    return nil;
}


@end

File: Example/VISPER/Example3/ExampleFeature3.swift
//
//  ExampleFeature3.swift
//  VISPER
//
//  Created by Jan Bartel on 09.03.16.
//  Copyright © 2016 Jan Bartel. All rights reserved.
//

import Foundation
import VISPER

class ExampleFeature3: VISPERFeature {

    override init() {
        super.init()
        self.addRoutePattern("/example3")
    }
    
    
    override func bootstrapWireframe(wireframe: IVISPERWireframe!, commandBus: VISPERCommandBus!) {
        super.bootstrapWireframe(wireframe, commandBus: commandBus)
        
        let handler = Example3CommandHandler()
        commandBus.addHandler(handler)
        
    }
    
    override func controllerForRoute(routePattern: String!, routingOptions options: IVISPERRoutingOption!, withParameters parameters: [NSObject : AnyObject]!) -> UIViewController! {
        
        if(routePattern == "/example3"){
            
            let presenter = Example3VisperViewControllerPresenter(wireframe: self.wireframe, commandBus: self.commandBus)
            
            let controller = Example3VisperViewController(nibName: "Example3VisperViewController", bundle: nil)
            
            controller.addVisperPresenter(presenter)
            
            return controller
        }
        
        return nil
    }
    
    
    override func optionForRoutePattern(routePattern: String!, parameters dictionary: [NSObject : AnyObject]?, currentOptions: IVISPERRoutingOption!) -> IVISPERRoutingOption? {
        
        if(currentOptions == nil && routePattern == "/example3"){
            return VISPER.routingOptionModal()
        }
        
        return currentOptions
    }
    
}
